### PR TITLE
Fix autopilot not working as expected on touch devices

### LINKED ISSUE #12731: Autopilot on touch platforms doesn't count clicks unless they are correctly positioned
**Describe the new feature:** 
Now, to interact with objects, I need to hover over the object itself, which is contrary to the essence of the mod.

**Proposal designs of the feature:** 
Is it possible to register clicks from the entire screen? I just want to follow the rhythm.

### PR DESCRIPTION
**Pull request by peppy**
Closes https://github.com/ppy/osu/issues/12731.

I haven't tested this, but quite confident it should work. Will test later today unless someone else beats me.

See
https://github.com/ppy/osu-framework/blob/dde9eb08ede4aa94d21590f92b7c60f85cac3947/osu.Framework/Input/InputManager.cs#L725-L733
for reference.

### CODE REVIEW

**Comment by peppy** [comment]
Tested to work on iOS.

File: osu.Game.Rulesets.Osu/OsuInputManager.cs [hunk 1]
```diff
@@ -39,6 +40,17 @@ protected override bool Handle(UIEvent e)
             return base.Handle(e);
         }
 
+        protected override bool HandleMouseTouchStateChange(TouchStateChangeEvent e)
+        {
+            if (!AllowUserCursorMovement)
+            {
+                // Still allow for forwarding of the "touch" part, but block the positional data.
+                e = new TouchStateChangeEvent(e.State, e.Input, e.Touch, false, null);
```
Comment (frenzibyte) [inline osu.Game.Rulesets.Osu/OsuInputManager.cs:48]: I think this makes more sense rather than forcing `IsActive` to false, which sounded wrong but apparently doesn't matter on base method:
```csharp
                var touch = new Touch(e.Touch.Source, CurrentState.Mouse.Position);
                e = new TouchStateChangeEvent(e.State, e.Input, touch, e.IsActive, null);
```

**Review by frenzibyte — COMMENTED** [review]
While this solves the issue, the code in `OsuInputManager.Handle` should've been enough in stopping touch move events from propagating to the state of this input manager, but it turns out that the code of translating touch events to inputs in `PassThroughInputManager` isn't respecting that very well. (and the blame goes to me for that 😛)

`TouchDownEvent` and `TouchUpEvent` do not get blocked by `Handle`, which is correct, but that causes `PassThroughInputManager` to copy their current positions as well while doing so, as can be seen [here](https://github.com/ppy/osu-framework/blob/2207064090e233cd49744bd8a3c24d6e8c0fde8d/osu.Framework/Input/PassThroughInputManager.cs#L113-L115).

~~I have a solution for that which is by handling each event type separately without mixing other properties alongside it.~~ (turns out it's not quite simple). But I've also noticed that `PassThroughInputManager` also applies touch positions from parent on `Sync`, which will be the case when resuming player from a pause, as that interferes with the state of `UseParentInput` which triggers `Sync` when true.

---

In summary, solving the underlying issue behind this requires 2 framework-side changes. While this solution on the other hand works as well, and can be considered a workaround for them instead.

~~I've removed the `TouchMoveEvent` from `Handle`'s condition as it's redundant now~~, but I have one small suggestion with this.

**Comment by peppy** [comment]
I don't understand why it's redundant. Are move events never sent on their own?

**Comment by frenzibyte** [comment]
They are sent but they'll end up at `HandleMouseTouchStateChange` anyways, which would ignore them. But now that I think about it again, it'll no longer be redundant when osu! ruleset starts to handle touch input directly, which may be what you're thinking about here... will bring it back.

File: osu.Game.Rulesets.Osu/OsuInputManager.cs (hunk 1 above)
Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/OsuInputManager.cs:48]: Is there a reason this change hasn't been applied @peppy ?

Comment (peppy) [reply osu.Game.Rulesets.Osu/OsuInputManager.cs:48]: I missed it. I think it looks fine as long as you test it to work as expected. Not 100% sure on the transfer of position from the mouse like that.

Comment (frenzibyte) [reply osu.Game.Rulesets.Osu/OsuInputManager.cs:48]: Yeah well, it's the only way to make it work in that direction, and I think it's fine personally as you're passing the current state of the mouse position, ignoring the position passed by the touch.

**Comment by smoogipoo** [comment]
I can't test this so relying on someone else to review.

**Review by frenzibyte — APPROVED** [review]
Tested with the latest changes, working as expected. 

Worthy to note that the gameplay cursor is incorrectly hidden in autopilot mod when playing in touchscreen, it's not easily fixable here as its state is managed at a `MenuCursorContainer` level, far away from the ruleset input managers. Will open an issue tracking that.
